patch tests for dynamics

btalamini opened the pull request:
Tests that solid mechanics physics module is able to exactly reproduce fields that are affine functions of space and time, which is essentially a test of consistency (in the sense of numerical approximation of differential equations).

Also implements a constant (in time) acceleration test, which tests second-order accuracy. This feature tests the body forcing term.

Bugs fixed in this PR:

- Sign convention in body forces
- Passes the correct time value from the ODE solver to the residual function
- Switches the solid mechanics default ODE integrator from `AverageAcceleration`, which turns out to be some [HHT-like method](https://github.com/mfem/mfem/blob/c644f3d1744af600f1a90c03e730c321543b58c5/linalg/ode.hpp#L785-L796), to the standard Newmark average acceleration method. This may not qualify as a "bug", but I was unable to get `AverageAcceleration` to pass the patch test. I do not yet know if this is an error in the MFEM implementation or if we are calling it incorrectly.

From here, we can replace the MFEM Newmark implementation with our own if we wish. I think it's good to have these tests in place first to reduce the chance of a regression.

This partially addresses #754. I'd leave that issue open until we make patch tests for `HeatTransfer`.

btalamini committed:
Get one dynamic patch test working, needs clean up

btalamini committed:
Get traction to compile, still doesn't pass

btalamini committed:
Make time integrator to call residual operator at correct discrete time point

Use the current scheme of passing references to variables around. Only
connected to solid mechanics modules for now.

btalamini committed:
Fix sign error in body force

btalamini committed:
Add debugging output

btalamini committed:
Set initial velocity correctly

btalamini committed:
Write out resultant force for debugging

btalamini committed:
Include inertial forces in reactions

Not sure if this is the right way to do it yet.

btalamini committed:
Restrict traction to essential boundaries to make reactions correct

btalamini committed:
Switch time integration scheme to Newmark, which seems to work

The default AverageAcceleration integrator is not getting the
right answer. The name is confusing - it is a variation of the HHT
integrator, not the classic Newmark method that goes by the
"average acceleration" name.

I don't know yet if the problem with AverageAcceleration is a bug
in its implementation, or if we are calling it incorrectly.

btalamini committed:
Go back to applying traction on all surfaces as it works for both 2D and 3D

Until we have a way to paint the boundaries, there is no good way to
restrict the traction application to only the complement of the
essential boundary condition surfaces. I've left in the old code
that does this filtering by the normal direction, so that it can
be resurrected easily to get reaction forces for debugging. I
could have a really ugly conditional for 2D and 3D. I'd rather not,
since it is brittle in that it depends on the particular mesh file
and the choice of essential/natural BC surfaces.

btalamini committed:
Rename to give reaction force operations a more obvious name

btalamini committed:
Switch default time integrator to Newmark, since it passes patch tests

btalamini committed:
Restore initial displacement in exact solution to make it more general

btalamini committed:
Clean up initializatino of exact solution to make it easier to read

btalamini committed:
Reorder for readability and reinstate disabled tests

btalamini committed:
Cleaning up to prepare for PR. No functionality change

btalamini committed:
Restore reaction calculation to original form. I want to tackle this in a separate PR

btalamini committed:
Merge branch 'develop' into testing/talamini/dynamic_patch_test2

btalamini committed:
Adapt to changes from main branch

btalamini committed:
Remove superseded tests

These legacy tests can be removed now that there are patch tests
covering the dynamics cases.

btalamini committed:
Fix test I broke in changing ODE interface

btalamini committed:
Integrate through multiple time steps to make test more thorough

btalamini committed:
Add constant acceleration test to check second order accuracy of Newmark

btalamini committed:
Merge branch 'develop' into testing/talamini/dynamic_patch_test2

btalamini committed:
Make style

jandrej commented:
Randomly saw this. `mfem::SecondOrderODESolver` is mostly user contributed. No exhaustive tests are done.

jamiebramwell commented:
Good to know @jandrej ! In general, we think having hooks to the fancier integrators is probably necessary for stiff systems like electrochemistry, but might be overkill for things like solid mechanics.

samuelpmishLLNL commented on src/serac/physics/tests/solid_dynamics_patch.cpp:
```diff
@@ -0,0 +1,438 @@
+enum class PatchBoundaryCondition
+{
+  Essential,
+  Mixed_essential_and_natural
```
The second enum value name format looks strange to me, it looks like most of serac's `enum class` definitions are using `PascalCase` for their values. It looks like I'm guilty of writing the most enums that don't follow this pattern, so I should go and fix those too!

btalamini replied:
Thanks, just updated it. Somewhat relatedly, this type is duplicated in the statics patch tests. We should consider factoring out some of these utilities out into test fixtures.

btalamini committed:
Fix minus sign syntax error

btalamini committed:
Fix capitalization convention of an enumeration

samuelpmishLLNL commented on src/serac/physics/tests/solid_dynamics_patch.cpp:
```diff
@@ -0,0 +1,438 @@
+      // The following matches the case when the top and left surfaces
+      // have essential boundary conditions:
+      //
+      // auto T = (n0[0] > 0.99 || n0[1] < -0.99)? dot(P, n0) : 0.0*n0;
```
some of these predicates might be easier to specify in terms of the spatial coordinates, rather than normals
```cpp
auto T = (x[0] < x[1]) * dot(P, n0);
```

btalamini replied:
That is true. But really, neither of those are the right way. Since this is just commented debug output, I prefer to leave it until we replace it with submesh boundary sets.

jamiebramwell replied:
Speaking of, we should have all the pieces to do that now.

white238 commented on src/serac/numerics/odes.hpp:
```diff
@@ -49,6 +49,8 @@ class SecondOrderODE : public mfem::SecondOrderTimeDependentOperator {
    * used by the residual operator
    */
   struct State {
+    double& time;
```
doxygen comment required

btalamini replied:
While I'm in there, I'll edit the doxygen comments on the other members that are misleading or incorrect.

samuelpmishLLNL commented on src/serac/physics/solid_legacy.hpp:
```diff
@@ -581,6 +581,8 @@ class SolidLegacy : public BasePhysics {
    */
   mfem::Vector previous_;
 
+  double ode_time_point_;
```
```suggestion
  /// @brief The value of time at which the ODE solver wants to evaluate the residual
  double ode_time_point_;
```

btalamini replied:
I moved this to the `BasePhysics` class and added the doxygen comment.

white238 commented on src/serac/physics/solid_legacy.cpp:
```diff
@@ -38,7 +38,8 @@ SolidLegacy::SolidLegacy(int order, const SolverOptions& options, GeometricNonli
       keep_deformation_(keep_deformation),
-      ode2_(displacement_.space().TrueVSize(), {.c0 = c0_, .c1 = c1_, .u = u_, .du_dt = du_dt_, .d2u_dt2 = previous_},
+      ode2_(displacement_.space().TrueVSize(),
+            {.time = ode_time_point_, .c0 = c0_, .c1 = c1_, .u = u_, .du_dt = du_dt_, .d2u_dt2 = previous_},
```
I can't find where `ode_time_point_` is getting initialized.

btalamini replied:
As per Sam's suggestion I moved `ode_time_point_` to the `BasePhysics` module. It is now initialized there.

white238 commented on src/serac/physics/solid_mechanics.hpp:
```diff
@@ -132,10 +132,12 @@ class SolidMechanics<order, dim, Parameters<parameter_space...>, std::integer_se
         shape_sensitivity_(mesh_, displacement_.space(), "shape_sensitivity"),
-        ode2_(displacement_.space().TrueVSize(), {.c0 = c0_, .c1 = c1_, .u = u_, .du_dt = du_dt_, .d2u_dt2 = previous_},
+        ode2_(displacement_.space().TrueVSize(),
+              {.time = ode_time_point_, .c0 = c0_, .c1 = c1_, .u = u_, .du_dt = du_dt_, .d2u_dt2 = previous_},
```
Wouldn't all these member variables be uninitialized at this point? I realize this was an existing problem...

btalamini committed:
Clarify doxygen comments for second order ODE solver

btalamini committed:
Put ODE solver time in base physics, since all modules will use it

btalamini committed:
Move function definition before its calls

Don't know how this compiled on my system, but it fails on CI.

btalamini committed:
Fix time integration test I broke by not reading carefully

This is maybe evidence that the system of storing global references
to send and receive data from the ODE integrator is confusing.

btalamini committed:
Fix style

jamiebramwell commented on src/serac/numerics/odes.cpp:
```diff
@@ -150,6 +150,7 @@ void SecondOrderODE::Solve(const double time, const double c0, const double c1,
 {
   // assign these values to variables with greater scope,
   // so that the residual operator can see them
+  state_.time  = time;
```
We should re-evaluate this interface strategy for ODE solvers. Tracking down the references and ownership of data can get confusing.

jamiebramwell reviewed (approved):
Thank you for tracking down the bugs!

jamiebramwell committed:
Merge branch 'develop' into testing/talamini/dynamic_patch_test2
